src/lib/timing.c

```diff
@@ -0,0 +1,231 @@
+		}
+#endif
+		if (clock_gettime(CLOCK_REALTIME, &ts) == 0) {
+			return (uint64_t)ts.tv_sec * NANOS_IN_SEC + (uint64_t)ts.tv_nsec;
+		}
+	}
+#endif
+
+#if HAVE_GETTIMEOFDAY | PHP_WIN32
+	// Fallback to gettimeofday() if better platform specific time is not
+	// available. gettimeofday() is always available in PHP on Windows, as it's
+	// expose through 'win32/time.[ch]'
+	{
+		struct timeval tp;
+
+		if (gettimeofday(&tp, NULL) == 0) {
+			return (uint64_t)tp.tv_sec * NANOS_IN_SEC + (uint64_t)tp.tv_usec * NANOS_IN_MICROSEC;
+		}
+	}
+#endif
+
+	// We give up
+	php_error(E_WARNING, "Xdebug could not determine a suitable clock source on your system");
+	return 0;
+}
+
+#if PHP_WIN32
+static uint64_t xdebug_counter_and_freq_to_nanotime(uint64_t counter, uint64_t freq)
+{
+	uint32_t mul = 1, freq32;
+	uint64_t q, r;
+
+	while (freq >= (1ULL << 32)) {
+		freq /= 2;
+		mul *= 2;
+	}
+	freq32 = (uint32_t)freq;
+
+	q = counter / freq32;
+	r = counter % freq32;
+	return (q * NANOS_IN_SEC + (r * NANOS_IN_SEC) / freq32) * mul;
+}
+#endif
+
+
+// Windows Windows 7 and lower
+#if PHP_WIN32
+static uint64_t xdebug_get_nanotime_rel(xdebug_nanotime_context *nanotime_context)
+{
+	LARGE_INTEGER tcounter;
+
+	if (nanotime_context->win_precise_time_func == NULL) {
+		QueryPerformanceCounter(&tcounter);
+		return xdebug_counter_and_freq_to_nanotime((uint64_t)tcounter.QuadPart, nanotime_context->win_freq);
+	}
+}
+#endif
+
+// Mac
+// should be fast but can be relative
+#if __APPLE__
+static uint64_t xdebug_get_nanotime_rel(xdebug_nanotime_context *nanotime_context)
+{
+	return clock_gettime_nsec_np(CLOCK_UPTIME_RAW);
+}
+#endif
+
+
+void xdebug_nanotime_init(void)
+{
+	xdebug_nanotime_context context = {0};
+
+#if PHP_WIN32
+	LARGE_INTEGER tcounter;
+
+	if (IsWindows8OrGreater()) {
+		context.win_precise_time_func = (WIN_PRECISE_TIME_FUNC)GetProcAddress(
+			GetModuleHandle(TEXT("kernel32.dll")),
+			"GetSystemTimePreciseAsFileTime"
+		);
+	} else {
+		context.win_precise_time_func = NULL;
+		QueryPerformanceFrequency(&tcounter);
+		context.win_freq = (uint64_t)tcounter.QuadPart;
+		context.use_rel_time = 1;
+	}
+#endif
+#if __APPLE__
+	context.use_rel_time = 1;
+#endif
+	context.start_abs = xdebug_get_nanotime_abs(&context);
+	context.last_abs = 0;
+#if PHP_WIN32 | __APPLE__
+	context.start_rel = xdebug_get_nanotime_rel(&context);
+	context.last_rel = 0;
+#endif
+
+	XG_BASE(nanotime_context) = context;
+}
+
+uint64_t xdebug_get_nanotime(void)
+{
+	uint64_t nanotime;
+	xdebug_nanotime_context *context;
+
+	context = &XG_BASE(nanotime_context);
+
+#if PHP_WIN32 | __APPLE__
+	/* Relative timing */
+	if (context->use_rel_time) {
+		nanotime = xdebug_get_nanotime_rel(context);
+
+		if (nanotime < context->last_rel + NANOTIME_MIN_STEP) {
+			context->last_rel += NANOTIME_MIN_STEP;
+			nanotime = context->last_rel;
+		}
+		context->last_rel = nanotime;
+		nanotime = context->start_abs + (xdebug_get_nanotime_rel(context) - context->start_rel);
```

Comment:
this seems like a mistake, why `xdebug_get_nanotime_rel` for 2nd time? (should be `nanotime` I think)